src/allmydata/test/mutable/test_version.py

```diff
@@ -47,343 +40,268 @@ def setUp(self):
+        self.version2_index = self.version2_seqnum - 1
+
+        results = await self._fn.download_version(self.servermap, self.version1)
+        self.assertThat(self.CONTENTS[self.version1_index],
+                        Equals(results))
+        results = await self._fn.download_version(self.servermap, self.version2)
+        self.assertThat(self.CONTENTS[self.version2_index],
+                        Equals(results))
+
+
+    async def test_download_nonexistent_version(self) -> None:
+        await self.do_upload_mdmf()
+        servermap = await self.mdmf_node.get_servermap(mode=MODE_WRITE)
+        await self.shouldFail(UnrecoverableFileError, "nonexistent version",
+                              None,
+                              self.mdmf_node.download_version, servermap,
+                              "not a version")
+
+
+    async def _test_partial_read(self, node, expected, modes, step) -> None:
+        version = await node.get_best_readable_version()
         for (name, offset, length) in modes:
-            d.addCallback(self._do_partial_read, name, expected, offset, length)
+            await self._do_partial_read(version, name, expected, offset, length)
         # then read the whole thing, but only a few bytes at a time, and see
         # that the results are what we expect.
-        def _read_data(version):
-            c = consumer.MemoryConsumer()
-            d2 = defer.succeed(None)
-            for i in range(0, len(expected), step):
-                d2.addCallback(lambda ignored, i=i: version.read(c, i, step))
-            d2.addCallback(lambda ignored:
-                self.assertThat(expected, Equals(b"".join(c.chunks))))
-            return d2
-        d.addCallback(_read_data)
-        return d
-
-    def _do_partial_read(self, version, name, expected, offset, length):
         c = consumer.MemoryConsumer()
-        d = version.read(c, offset, length)
+        for i in range(0, len(expected), step):
+            await version.read(c, i, step)
+        self.assertThat(expected, Equals(b"".join(c.chunks)))
+
+    async def _do_partial_read(self, version, name, expected, offset, length) -> None:
+        c = consumer.MemoryConsumer()
+        await version.read(c, offset, length)
         if length is None:
             expected_range = expected[offset:]
         else:
             expected_range = expected[offset:offset+length]
-        d.addCallback(lambda ignored: b"".join(c.chunks))
-        def _check(results):
-            if results != expected_range:
-                print("read([%d]+%s) got %d bytes, not %d" % \
-                      (offset, length, len(results), len(expected_range)))
-                print("got: %s ... %s" % (results[:20], results[-20:]))
-                print("exp: %s ... %s" % (expected_range[:20], expected_range[-20:]))
-                self.fail("results[%s] != expected_range" % name)
-            return version # daisy-chained to next call
```

Comment:
`_do_partial_read()` used to return the `version` and now it returns None?

Reply:
But I guess that's not necessary anymore, it can just handle original version directly?

Reply:
Correct.  The daisy-chaining was to deal with the different callback/test function scopes.  With `await` and no callbacks, there's no extra scopes so there's no need to propagate this value along a callback chain to make it available later on.  It's just available in the test method scope.